osu.Game.Rulesets.Osu/Objects/Drawables/SpinnerSpinHistory.cs

```diff
@@ -0,0 +1,179 @@
+        public void RemoveDelta(double currentTime, float delta)
+        {
+            while (currentTime < currentTurn.StartTime)
+            {
+                // When crossing over a turn, we need to adjust the delta so that it's relative to the end point of the next turn.
+                //
+                // This is done by ADDING the delta between the current turn and the next turn.
+                // To understand why this is, notice that delta is a rate-independent value. Suppose the turn values are { 90, 45 } (i.e. CW then CCW spin)...
+                // - If delta < 0 (e.g. -15) (i.e. CCW rotation), then the next turn should be <45 (therefore delta = -15 + (45 - 90) = -60, next = 30).
+                // - If delta = 0, then the next turn should be =45 (therefore delta = 0 + (45 - 90) = -45, next = 45).
+                // - If delta > 0 (e.g. +15) (i.e. CW rotation), then the next turn should be >45 (therefore delta = 15 + (45 - 90) = -30, next = 60).
+                //
+                // There is a special case when crossing a complete spin, because the turn following it starts at 0 rather than the previous turn's value.
+                // In this case, only the remaining delta in the current turn needs to be considered.
+
+                Turn nextTurn = turningPoints.Pop();
+
+                if (nextTurn.IsCompleteSpin)
+                    delta += currentTurn.Angle;
+                else
+                    delta += currentTurn.Angle - nextTurn.Angle;
+
+                currentTurn = nextTurn;
+            }
+
+            currentTurn.Angle += delta;
+
+            // Note: Enumerating through a stack is already reverse order.
+            currentMaxRotation = turningPoints.Prepend(currentTurn).TakeWhile(t => !t.IsCompleteSpin).Select(t => Math.Abs(t.Angle)).Max();
+            TotalRotation = 360 * turningPoints.Count(t => t.IsCompleteSpin) + currentMaxRotation;
+        }
+
+        /// <summary>
+        /// Finishes the current turn and starts a new one from its end point.
+        /// </summary>
+        /// <param name="currentTime">The start time of the new turn.</param>
+        /// <param name="direction">The turning direction.</param>
+        private void beginNewTurn(double currentTime, int direction)
+        {
+            turningPoints.Push(currentTurn);
+            currentTurn = new Turn(currentTime, direction) { Angle = currentTurn.Angle };
+        }
+
+        /// <summary>
+        /// Represents a single direction turn of the spinner.
+        /// </summary>
+        private struct Turn
+        {
+            /// <summary>
+            /// The start time of this turn.
+            /// </summary>
+            public readonly double StartTime;
+
+            /// <summary>
+            /// The turning direction.
+            /// </summary>
+            public readonly int Direction;
+
+            /// <summary>
+            /// The current angle.
```

Comment:
```suggestion
            /// The current absolute angle.
```
i guess

Reply:
So to clarify, this "Turn" class is not storing anything but absolute angles right? There's no delta stored here.

Maybe naming it `TurningPoint` is better if that's the case?